### PR TITLE
feat: add `CharmReconciler` and `Components`

### PR DESCRIPTION
**Pull request by ca-scribner**
This implements:
* a reusable charm reconcile function (`CharmReconciler`)
* `Component`, an abstraction for any piece of logic in a charm
* several generic `Component` subclasses for Kubernetes resources, Pebble containers, and some other situations

These are documented in:
* `./src/charmed_kubeflow_chisme/components/README.md`, which gives a general overview
* `./src/charmed_kubeflow_chisme/components/how-to/`, has several how-to guides for using these tools
* [this spec](https://docs.google.com/document/d/1AOmp25BfyEWaPlSuS3U9UlF_Q5CKMTwHw8T84_KwDEM/edit)

Tests are included, but during development this was also tested live with the following prototype charms:
* https://github.com/canonical/kfp-operators/pull/264
* https://github.com/canonical/kfp-operators/pull/268
* https://github.com/canonical/kfp-operators/pull/280
* https://github.com/canonical/kfp-operators/pull/287
* https://github.com/canonical/kfp-operators/pull/288

### CODE REVIEW

File: src/charmed_kubeflow_chisme/status_handling/multistatus.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    def add(self, component: str, get_status: typing.Callable[[], ops.StatusBase]):
+        """Add a named status component."""
+        if component in self._components:
+            raise ValueError(f"duplicate component {component!r}")
+        self._components[component] = get_status
+
+    def highest(
+        self, statuses: Optional[List[Tuple[str, ops.StatusBase]]] = None
+    ) -> ops.StatusBase:
+        """Return highest-priority status with message prefixed with component name."""
```
Comment (i-chvets) [inline src/charmed_kubeflow_chisme/status_handling/multistatus.py:42]: Will we be able to see all other status/error/info messages in the log?

File: tests/unit/components/fixtures.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+from charmed_kubeflow_chisme.components.component_graph_item import ComponentGraphItem
+from charmed_kubeflow_chisme.components.pebble_component import (
+    PebbleComponent,
+    PebbleServiceComponent,
+)
+
+COMPONENT_NAME = "component"
+
+
+class MinimallyExtendedComponent(Component):
```
Comment (i-chvets) [inline tests/unit/components/fixtures.py:19]: Docstrings in methods in this class.

File: tests/unit/components/test_charm_reconciler.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        # Arrange
+        charm = harness.charm
+
+        charm_reconciler = CharmReconciler(charm)
+        charm_reconciler.install_default_event_handlers()
+        with pytest.raises(RuntimeError):
+            charm_reconciler.install_default_event_handlers()
+
+
+class TestUpdateStatus:
```
Comment (i-chvets) [inline tests/unit/components/test_charm_reconciler.py:88]: Docstrings in methods in this class.

File: tests/unit/components/test_component_graph.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+            # Components here of the same name, we will get a RuntimeError from the harness.
+            name = "component"
+
+        cg = ComponentGraph()
+        cg.add(component=MockComponent())
+        with pytest.raises(ValueError):
+            cg.add(component=MockComponent())
+
+
+class TestGetExecutableComponentItems:
```
Comment (i-chvets) [inline tests/unit/components/test_component_graph.py:58]: Docstrings in methods in this class.

File: tests/unit/components/test_component_graph.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        cgi2.component.configure_charm("mock event")
+
+        assert next(cgi_generator) == cgi4
+
+        # And now the generator should be empty
+        with pytest.raises(StopIteration):
+            next(cgi_generator)
+
+
+class TestEventsToObserve:
```
Comment (i-chvets) [inline tests/unit/components/test_component_graph.py:191]: Docstrings in methods in this class.

File: tests/unit/components/fixtures.py (hunk 2 above)
Comment (ca-scribner) [reply tests/unit/components/fixtures.py:19]: Done!

File: tests/unit/components/test_charm_reconciler.py (hunk 3 above)
Comment (ca-scribner) [reply tests/unit/components/test_charm_reconciler.py:88]: Done!

File: tests/unit/components/test_component_graph.py (hunk 4 above)
Comment (ca-scribner) [reply tests/unit/components/test_component_graph.py:58]: Done!

File: tests/unit/components/test_component_graph.py (hunk 5 above)
Comment (ca-scribner) [reply tests/unit/components/test_component_graph.py:196]: Done!

File: src/charmed_kubeflow_chisme/status_handling/multistatus.py (hunk 1 above)
Comment (ca-scribner) [reply src/charmed_kubeflow_chisme/status_handling/multistatus.py:42]: Not from `highest`, but the CharmReconciler.reconcile calls `all` first so it has all the statuses.  The `highest` goes to the actual charm status on `juju status`, but everything else is in `debug-log`

**Review by i-chvets — APPROVED** [review]
Approved.
This branch was tested with multiple sidecare re-writes.

**Pull request merged**